pkg/query-service/app/opamp/configure_ingestionRules.go

```diff
@@ -0,0 +1,228 @@
+	service := agentConf.Get("service")
+	signalSpec := service.(map[string]interface{})["pipelines"].(map[string]interface{})[signal]
+	exporters := signalSpec.(map[string]interface{})["exporters"].([]interface{})
+	var found bool
+	for _, e := range exporters {
+		if e == name {
+			found = true
+		}
+	}
+
+	return found
+}
+
+// addIngestionControlToAgent adds ingestion contorl rules to agent config
+func addIngestionControlToAgent(agent *model.Agent, signal string, processors map[string]interface{}, withLB bool) (string, error) {
+	confHash := ""
+	config := agent.EffectiveConfig
+	c, err := yaml.Parser().Unmarshal([]byte(config))
+	if err != nil {
+		return confHash, err
+	}
+
+	agentConf := confmap.NewFromStringMap(c)
+
+	// check if LB needs to be configured
+	if signal == string(Traces) && withLB {
+
+		// check if lb pipeline is setup for lb agents, here we assume
+		// that otlp_internal will exist implicitly if traces/lb pipeline exists
+		if agent.CanLB && !checkPipelineExists(agentConf, TracesLbPipelineName) {
+			zap.S().Debugf("LB not configured for agent. Preparing the lb exporter spec", agent.ID)
+			serviceConf, err := prepareLbAgentSpec(agentConf)
+			if err != nil {
+				zap.S().Error("failed to prepare lb exporter spec for agent", agent.ID, err)
+				return confHash, err
+			}
+			if err := agentConf.Merge(serviceConf); err != nil {
+				zap.S().Error("failed to merge lb exporter spec for agent", agent.ID, err)
+				return confHash, err
+			}
+		}
+
+		// check if otlp_internal is setup for non-lb agents
+		if !agent.CanLB && !checkExporterExists(agentConf, string(Traces), OtlpInternalReceiver) {
+			zap.S().Debugf("OTLP internal not configured for agent. Preparing the lb exporter spec", agent.ID)
+			serviceConf, err := prepareNonLbAgentSpec(agentConf)
+			if err != nil {
+				zap.S().Error("failed to prepare OTLP internal receiver spec for agent", agent.ID, err)
+				return confHash, err
+			}
+			if err := agentConf.Merge(serviceConf); err != nil {
+				zap.S().Error("failed to merge OTLP internal receiver spec for agent", agent.ID, err)
+				return confHash, err
+			}
+		}
+	}
+
+	// add ingestion control
+	serviceConf, err := prepareControlProcesorsSpec(agentConf, Signal(signal), processors)
+	if err != nil {
+		zap.S().Error("failed to prepare ingestion control processors for agent ", agent.ID, err)
+		return confHash, err
+	}
+
+	err = agentConf.Merge(serviceConf)
+	if err != nil {
+		return confHash, err
+	}
+
+	// ------ complete adding processor
+	configR, err := yaml.Parser().Marshal(agentConf.ToStringMap())
+	if err != nil {
+		return confHash, err
+	}
+
+	zap.S().Debugf("sending new config", string(configR))
+	hash := sha256.New()
+	_, err = hash.Write(configR)
+	if err != nil {
+		return confHash, err
+	}
+	confHash = string(hash.Sum(nil))
+	agent.EffectiveConfig = string(configR)
+	err = agent.Upsert()
+	if err != nil {
+		return confHash, err
+	}
+
+	agent.SendToAgent(&protobufs.ServerToAgent{
```

Comment:
I don't think this should send the status or interact with each agent directly. There is one already on the `Server` why this again?

Reply:
do you mean onMessage()? I am not sure this was implemented by Nitya. Can you share which method to be used instead of this to interact with agent?

Reply:
Not `onMessage`, the server also has `UpsertProcessor` function added in this PR which I believe is supposed to do the same.

Reply:
UpsertProcessor() is being used for log pipelines which Nitya is working on. He wanted a different level of control on how processors are inserted.

For ingestion rules, I have UpsertIngestionControlProcs() and for log pipelines there is UpsertProcessor().
Both can be separate in this release as it helps us move in different direction (me and Nitya). We can merge them in future versions once this whole things becomes stable.

Reply:
yeh. agree. I will include this in separate PR tomorrow.  Added issue for all improvements[here](https://github.com/SigNoz/engineering-pod/issues/963)